File: signature-suite-impls/java/signaturesuite/src/main/java/did/Did.java
package did;

import canonical.Canonical;
import net.i2p.crypto.eddsa.EdDSAPublicKey;
import net.i2p.crypto.eddsa.spec.EdDSAPrivateKeySpec;
import net.i2p.crypto.eddsa.spec.EdDSAPublicKeySpec;
import org.bitcoinj.core.Base58;
import proof.Proof;
import proof.Provable;

import java.util.Arrays;
import java.util.UUID;

public final class Did
        extends UnsignedDidDoc
        implements Provable {
    public static final String INITIAL_KEY = "key-1";
    public static final String DID_METHOD_PREFIX = "did:work:";

    private Proof proof;

    private Did(final UnsignedDidDoc unsignedDidDoc, final Proof proof) {
        super(unsignedDidDoc.getId(), unsignedDidDoc.getPublicKey(), unsignedDidDoc.getAuthentication(),
                unsignedDidDoc.getService());
        this.proof = proof;
    }

    public static String generateDID(final EdDSAPublicKey pubKey) {
        final byte[] pubKeyBytes = pubKey.getAbyte();
        final byte[] firstSixteen = Arrays.copyOfRange(pubKeyBytes, 0, 16);
        return DID_METHOD_PREFIX + Base58.encode(firstSixteen);
    }

    public static Did signDIDDoc(final UnsignedDidDoc unsignedDidDoc,
                                 final EdDSAPrivateKeySpec privKey,
                                 final String keyRef)
            throws Exception {
        final String nonce = UUID.randomUUID().toString();
        Did didDoc = new Did(unsignedDidDoc, null);
        final Proof proof = Proof.createEd25519Proof(didDoc, privKey, keyRef, nonce);
        didDoc.setProof(proof);
        return didDoc;
    }

    public static boolean validateDidDocProof(final Did didDoc, final EdDSAPublicKeySpec pubKey)
            throws Exception {
        return Proof.verifyEd25519Proof(didDoc, pubKey);
    }

    public UnsignedDidDoc getUnsignedDidDoc() {
        return new UnsignedDidDoc(this.getId(), this.getPublicKey(), this.getAuthentication(), this.getService());
    }

    public Proof getProof() {
        return proof;
    }

    @Override public void setProof(Proof proof) {
        this.proof = proof;
    }

    @Override public String toJson() {
        return Canonical.toJson(this);
    }

    @Override public String toString() {
        return Canonical.toJson(this);
    }
}

File: signature-suite-impls/golang/did/did.go
package did

import (
	"github.com/google/uuid"
	"github.com/mr-tron/base58"
	"golang.org/x/crypto/ed25519"

	"signaturesuite/proof"
)

type UnsignedDIDDoc struct {
	ID             string       `json:"id"`
	PublicKey      []KeyDef     `json:"publicKey"`
	Authentication []string     `json:"authentication"`
	Service        []ServiceDef `json:"service"`
}

// DIDDoc a W3C compliant signed DID Document
type DIDDoc struct {
	*UnsignedDIDDoc
	*proof.Proof `json:"proof"`
}

func (d *DIDDoc) GetProof() *proof.Proof {
	return d.Proof
}

func (d *DIDDoc) SetProof(p proof.Proof) {
	d.Proof = &p
}

// KeyDef represents a DID public key
type KeyDef struct {
	ID              string `json:"id"`
	Type            string `json:"type"`
	Controller      string `json:"controller,omitempty"`
	PublicKeyBase58 string `json:"publicKeyBase58"`
}

type ServiceDef struct {
	ID              string `json:"id"`
	Type            string `json:"type"`
	ServiceEndpoint string `json:"serviceEndpoint"`
}

const (
	// InitialKey the key reference assigned to the first key in a DidDoc
	InitialKey      = "key-1"
	IssuerDIDMethod = "did:work:"
)

// GenerateDID generate a DID
func GenerateDID(publicKey *ed25519.PublicKey) string {
	bytes := *publicKey
	return IssuerDIDMethod + base58.Encode(bytes[0:16])
}

func SignDIDDoc(unsignedDoc UnsignedDIDDoc, privKey ed25519.PrivateKey, keyRef string) (*DIDDoc, error) {
	doc := DIDDoc{
		UnsignedDIDDoc: &unsignedDoc,
		Proof:          nil,
	}
	nonce := uuid.New().String()
	docProof, err := proof.CreateEd25519Proof(&doc, privKey, keyRef, nonce)
	doc.Proof = docProof
	return &doc, err
}

func ValidateDIDDocProof(didDoc DIDDoc, pubKey ed25519.PublicKey) error {
	return proof.VerifyEd25519Proof(&didDoc, pubKey)
}
